articles/141_static_remapping.md

```diff
@@ -0,0 +1,342 @@
+It works by first expanding the relative name and then doing FQN replacement.
+This is also part of ROS 1 remapping.
+
+*Example:*
+
+- A node is in namespace `/ns/` and uses name `bar`
+- A rule is created to remap `bar` to `foo`
+- The node's use of `bar` is expanded to `/ns/bar`
+- Both sides of the remap rule are expanded to `/ns/bar` and `/ns/foo`
+- Because the FQN match, the final name is `/ns/foo`
+
+
+### Remap via Command Line
+A user can supply node specific remapping arguments via the command line.
+Because a process can contain multiple nodes, there must be a way to uniquely identify a node in a process.
+This is a feature of ROS 1 remapping.
+
+*Example:*
+
+- Node is an executable at `/bin/my_node`
+- User wants to change `/cat` to `/dog`
+- User types `/bin/my_node /cat:=/dog`
+
+
+### Change the Default Namespace
+The default namespace is the one in which relative names get expanded to.
+This should be changeable without affecting FQN.
+ROS 1 has this feature using either the environment variable `ROS_NAMESPACE` or the argument `__ns`.
+
+*Example:*
+
+- Node is in namespace `/ns` and uses relative name `bar`
+- User changes the default namespace to `/foo`
+- The final name is `/foo/bar`
+
+## Remapping Names in ROS 1
+Remapping is a feature that also exists in ROS 1.
+In ROS 1 remapping works by passing in [arguments](http://wiki.ros.org/Remapping%20Arguments) to each node.
+Client libraries also have APIs in code to pass remapping rules when the node is initialized.
+A remap rule consists of two names: one that should be replaced with another.
+
+ROS 1 remapping works on **Fully Qualified Names** (FQN).
+Both sides of a rule are [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#a377ff8fede7b95398fd5d1c5cd49246b).
+Before a name is remapped it is also [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#ab2eebaf734abfbdccb4122f8e24f547f).
+The name is remapped to the right side only if it exactly matches the left side of a rule.
+
+
+## Remapping rule syntax
+This is a proposal for the ROS 2 remapping rule syntax.
+It attempts to be the same as ROS 1 syntax when possible, and different where it supports new features.
+
+Requirements supported by this syntax
+- Remap One Node in a Process
+- Change a Namespace
+- Change a Basename
+- Exact FQN Replacement
+- Exact Relative Name Replacement
+- Remap via Command Line
+- Change the Default Namespace
+
+Requirements not supported
+- Change a Token
+- Pre-FQN Remapping
+
+Special Operators
+- `*` wild card that matches a single token
+- `**` wild card that matches at least one slash and zero or more tokens
+- `:=` divides the two parts of a remapping rule: matching and replacement
+- `~` expands to `/namespace/nodename` and may only be at the very beginning of a match or replacement
+- `__ns` matches the default namespace if used by itself on the match side of a rule
```

Comment:
Do you think this is sufficient? I'm thinking of a case where the remappings are calculated in order, and you have one normal remap, then a remap that starts with `__ns:=`, then another. In that case the left hand side (the "match") would be affected in the last remapping by the middle one. I don't that happens with other remappings. That's what makes it "special" in my mind.

I don't know how it works in ROS 1 or if it even matters in that case, but originally I was thinking that changing the node's name or it's namespace (`__name:=` and `__ns:=` respectively) might need to be handled before all other remappings. That might evolve if we allow either or both of those to be changed during runtime. Perhaps it's worth distinguishing between changing the namespace with `__ns:=` and a normal namespace remap whose "match" part is just the namespace of the node. The former would potentially be retroactive, whereas the latter might be applied in order like the other remap rules.

Not sure if I have a concrete change suggestion, but I'd like to know what you think about this point.